subprojects/core/src/main/java/org/gradle/api/internal/TaskOutputsInternal.java

```diff
@@ -47,9 +49,8 @@
     boolean hasDeclaredOutputs();
 
     /**
-     * Returns {@code false} if the task declares any multiple-output properties via {@link #files(Object...)},
-     * {@literal @}{@link org.gradle.api.tasks.OutputFiles} or
-     * {@literal @}{@link org.gradle.api.tasks.OutputDirectories}; or {@code true} otherwise.
+     * Returns whether the outputs can be cached.
+     * @param reasons Reasons for disabling the cache are added to this list.
      */
-    boolean isCacheAllowed();
+    boolean isCacheEnabled(List<TaskCachingDisabledReason> reasons);
```

Comment:
At some point it felt like I wanted to distinguish between the task not being annotated at all and some other reasons like using `doNotCacheIf`. Looking at this closer I think we should be fine with having just a list of Strings. It feels like a good idea to expose the interface `TaskOutputCaching` to the listener? The other option would be to not use a listener at all but to have the client call `task.getOutputs().getTaskOutputCaching()`. WDYT?